### ISSUE TITLE
Support for Arabic or other languages

### ISSUE THREAD

**Issue by xakiru** [rhasspy/rhasspy#183]
Is there a way to train a model in order to create other languages profiles, like arabic for instance?
Thanks you

**Comment by synesthesiam** [comment]
Yes, this is definitely possible. My approach would be to train a Kaldi ASR model using IPA phonemes. A challenge with Arabic looks to be the writing system. An extra library would be needed to intelligently add vowels back in.

**Comment by B0ndo2** [comment]
I would like to help with Arabic, am a native speaker and very much interested in getting STT in Arabic. How and where can I start ?
Looks like some work was done [here](https://medium.com/@omar.merghany95/how-we-built-arabic-speech-recognition-system-using-kaldi-a10a54678180)

**Comment by synesthesiam** [comment]
Hi @B0ndo2, thank you for the offer and the link.

It looks like the limiting factor for me is going to be audio data. I have close to 90 hours so far, which is a good start. Do you know of any free, large Arabic speech datasets besides Common Voice? I have about 20 hours from a handful of other datasets.

I can always try training the model and seeing how it does. Would you be willing to write up Arabic versions of the [default English sentences](https://github.com/rhasspy/rhasspy-profile/blob/master/rhasspyprofile/profiles/en/sentences.ini)?

**Comment by B0ndo2** [comment]
What does the dataset look like, I can help with searching. I am willing to write Arabic versions of the default English, how do I start ?

**Comment by synesthesiam** [comment]
The datasets are pretty simple: just audio files (WAV, MP3, etc.) with spoken sentences and the text of what was said (usually in a text or CSV file). The hard part is getting 100's of hours with many different speakers, and making sure licensing is OK (no commercial audiobooks).

>  I am willing to write Arabic versions of the default English, how do I start ?

Thanks! You'll need to know a little bit of the [template syntax](https://rhasspy.readthedocs.io/en/latest/training/). I just need some example intents and voice commands that a native Arabic speaker would actually use. For example, I have:

```ini
[GetTime]
what time is it
```

is translated to (according to Google Translate):

```ini
[GetTime]
أي ساعة
```

If I were to run all of the English sentences through Google Translate, I'm guessing a lot of them would sound wrong to a native speaker. Feel free to add your own intents (like `GetTime`) :slightly_smiling_face:

**Comment by B0ndo2** [comment]
I can translate the ones that exists and add some more,  once I have them
translated how do I commit them?
For the audio files, can I use news/radio and create the written text for
them (what about timing and sync between the text and the audio files?).
What is minimum number of hours to get started (in addition to those I
shared earlier)?

**Comment by synesthesiam** [comment]
> once I have them translated how do I commit them?

You can post them here or e-mail them to me, and I can add them to the `rhasspy-profile` project.

> For the audio files, can I use news/radio and create the written text for them (what about timing and sync between the text and the audio files?).

News/radio should be fine. I forgot to mention that you should also include a speaker id, or separate the audio/text files by speaker. Timing/sync is not needed unless you have the audio in one big file, which is fine too -- in that case, each sentence should be marked with start/end times, speaker id, and text.

> What is minimum number of hours to get started (in addition to those I shared earlier)?

This seems to depend heavily on the language, but 100 hours is usually a good start. For high accuracy, more than 1000 hours is needed.

**Comment by B0ndo2** [comment]
These are easy (This is Arabic for Egypt)
[GetTime]
كام الساعة
الساعة كام

[GetTemperature]
كام الحرارة
الحرارة كام

[GetGarageState]
باب الجراج مفتوح او مقفول

This one is challenging to me, not sure what gets translate to Arabic and what stays in English
[ChangeLightColor]
light_name = (bedroom light) {name} --> light_name = (نور غرفة النوم){name}
color = (red | green | blue) {color} -->(احمر| اخضر | ازرق){color}

set [the] <light_name> [to] <color> -->  غير لون   <light_name> الى <color>

100 hours is a good start, can you use the 90 hours from the project I shared with you ?.
When you say a speaker id, can I do this:
speaker_id1.mp3, speaker_id1.txt
speaker_id2.mp3, speaker_id2.txt
etc

**Comment by synesthesiam** [comment]
> This one is challenging to me, not sure what gets translate to Arabic and what stays in English

light_name, color, and the {name} part should be the only English. Everything else (set, the, to) should be Arabic. A [word] in brackets just means its optional.

> 100 hours is a good start, can you use the 90 hours from the project I shared with you ?.

I didn't see anywhere to download the data they mention. The [paper in their repository](file:///tmp/mozilla_hansenm0/Final-Report_2-1.pdf) says it was provided by "RDI Company". I'd be happy to include the data if it's publicly available and has an open license.

**Comment by B0ndo2** [comment]
I couldn't get a copy of the dataset from RDI (unfortunately they refused to share it or even license it). I came across [this](http://en.arabicspeechcorpus.com/) for Arabic TTS

**Comment by synesthesiam** [comment]
OK, so far I have 95 hours split across (1) Common Voice, (2) Arabic Speech Corpus, (3) Linga Libre, and (4) Tunisian MSA. I'll train a Kaldi model with this data and see how it turns out :slightly_smiling_face:

**Comment by B0ndo2** [comment]
Hi  Michael,
 How is the training going?



*Taymour*

**Comment by synesthesiam** [comment]
Training was successful, and reports a low word-error rate (7.7%). The test set is pretty small, though, so we'll have to see how it performs in the real world. I got it to recognize all of the `GetTime`, `GetTemperature`, and `GetGarageState` sentences you wrote (as spoken by Google Translate).

The biggest restriction right now is guessing the pronunciations of unknown words. I trained a PyTorch model to do this for training purposes, but it needs a GPU so I can't deploy it with Rhasspy. I usually use [phonetisaurus](https://github.com/AdolfVonKleist/Phonetisaurus) for this purpose, but it failed to train an Arabic model; I'm guessing this is because of the missing diacritics problem.

I'm looking into [another solution](https://github.com/Kyubyong/g2p) that only needs numpy to run (still trained with PyTorch).

**Comment by B0ndo2** [comment]
How can I help, do you want me to add the diacritics ?, speak the sentences
?




*Taymour*

**Comment by synesthesiam** [comment]
One way to help would be to tell me if the words below are good examples for each phoneme.
The format is `<PHONEME> <WORD> <PRONUNCIATION>`
I got the list of phonemes from [Wikipedia](https://en.wikipedia.org/wiki/Arabic_phonology), and there are pronunciations available at [ipachart](https://www.ipachart.com/)

The goal is to pick a word that a native speaker would know for sure contains the phoneme sound. In English, this would be like choosing "dog" for the "d" sound, or "cat" for "æ".

```
a آمنة aː m i n a
aj اللي aː l l aj j
aw الجو aː l d͡ʒ aw w
aː آثار aː θ aː r
b أيوب ʔ aj j uː b
d أحدا ʔ a ħ a d d aː
dˤ أيضا ʔ i i dˤ aː
dˤˤ الموضة aː l m uː a dˤˤ
d͡ʒ ارجو ʔ a r d͡ʒ uː
f آفاق aː f aː q
h أمره ʔ a m a r r a h u
i أخيه ʔ a x iː h i
iː أكيد ʔ a k iː d
j أيام ʔ aj j aː m
k أنكم ʔ a n k a m a
l آلات aː l aː t
m آلام aː l aː m
n أمين ʔ a m iː n
q أقول ʔ a q uː l
r آراء aː r aː ʔ
s أساس ʔ a s aː s
sˤ الصف aː l sˤ a f f
sˤˤ النص aː l n a sˤˤ
t آيات aː j aː t
tˤ إطار ʔ i tˤ aː r
tˤˤ الخط aː l x a tˤˤ
u أخرى ʔ u x r aː
uː أرجو ʔ a r d͡ʒ uː
w أسوأ ʔ a s w a ʔ
x أختي ʔ a x t iː
z أزمة ʔ a z a m a
ð أخذت ʔ a x ð u t
ðˤ ظروف ðˤ u r uː f
ðˤˤ لحظة l a ħ a ðˤˤ a
ħ إحدى ʔ i ħ d aː
ɣ الغد aː l ɣ a d
ʃ الشر aː l ʃ a r r
ʔ أئمة ʔ a ʔ m a
ʕ أشعة ʔ a ʃ ʕ a
θ بحيث b i ħ i i θ u
```

**Comment by B0ndo2** [comment]
Here is my version (based on Egyptian Arabic)

a آمنة aː m e n a -- > modified
aj اللي aː l l aj j --> not good
aw الجو aː l g aw w --> modified
aː آثار aː θ aː r
b أيوب ʔ aj j uː b
d أحدا ʔ a ħ a d d aː --> not good
dˤ أيضا ʔ a i dˤ aː n --> modified
dˤˤ الموضة aː l m uː a dˤˤ --> not good
d͡ʒ ارجو ʔ a r g uː --> modified
f آفاق aː f aː q
h أمره ʔ a m a r r a h u
i أخيه ʔ a x iː h i
iː أكيد ʔ a k iː d
j أيام ʔ aj j aː m
k أنكم ʔ a n n a k? o m a --> modified
l آلات aː l aː t --> not good
m آلام aː l aː m
n أمين ʔ aː m iː n --> modified
q أقول ʔ a q uː l
r آراء aː r aː ʔ
s أساس ʔ a s aː s
sˤ الصف aː l sˤ a f f
sˤˤ النص aː l n a sˤˤ
t آيات aː j aː t
tˤ إطار ʔ i tˤ aː r
tˤˤ الخط aː l x a tˤˤ
u أخرى ʔ u x r aː
uː أرجو ʔ a r g uː --> modified
w أسوأ ʔ a s w a ʔ
x أختي ʔ oː x t iː --> modified
z أزمة ʔ a z m a --> modified
ð أخذت ʔ a x a ð t u --> modified
ðˤ ظروف ðˤ u r uː f
ðˤˤ لحظة l a ħ a ðˤˤ a
ħ إحدى ʔ i ħ d aː
ɣ الغد aː l ɣ a d d --> modified
ʃ الشر aː l ʃ a r r
ʔ أئمة ʔ a ʔe m a --> modified
ʕ أشعة ʔ a ʃ eʕ a --> modified
θ بحيث b i ħ a i θ u --> modified



*Taymour*

**Comment by synesthesiam** [comment]
Thank you! Can you suggest replacements for the "not good" ones?

**Comment by B0ndo2** [comment]
How about these

d حديد ħ a d  i d
aj أيضا  aj dˤˤ a n
l لكن l a k? e: n

**Comment by synesthesiam** [comment]
Arabic support will hopefully be added to the next version of Rhasspy.

**Comment by B0ndo2** [comment]
I am available to test



*Taymour*

**Comment by synesthesiam** [comment]
OK, and thanks for your patience! I will contact you in the next few weeks. Are you able to use the Docker version of Rhasspy, or do you do something else?

**Comment by B0ndo2** [comment]
I can use the docker version, currently am using the .deb versions on a Pi




*Taymour*

**Comment by synesthesiam** [comment]
The latest Rhasspy preview has Arabic support build in. To get it, do:

`docker pull rhasspy/rhasspy:2.5.11` and then run as normal:

```sh
docker run -d -p 12101:12101 \
      -it \
      -v "$HOME/.config/rhasspy/profiles:/profiles" \
      -v "/etc/localtime:/etc/localtime:ro" \
      --device /dev/snd:/dev/snd \
      rhasspy/rhasspy \
      --user-profiles /profiles \
      --profile ar
```

**Comment by B0ndo2** [comment]
Is it also available as .deb ?




*Taymour*

**Comment by synesthesiam** [comment]
Not yet, but I can work on that. Are you on a regular desktop, or a Pi?

**Comment by B0ndo2** [comment]
My production is on a pi



*Taymour*

**Comment by synesthesiam** [comment]
Debian packages for 2.5.11 are finally available: https://github.com/rhasspy/rhasspy/releases/tag/v2.5.11

**Comment by B0ndo2** [comment]
Sorry it took me some time to test. I added "كم الساعة" to my sentences
files on Rhasspy addon in HA and training fails with the message below
[GetTime]
what time is it
tell me the time
كم الساعة

TrainingFailedException: Command '['fstcompile',
'--isymbols=/share/rhasspy/profiles/en/kaldi/model/data/lang/words.txt',
'--osymbols=/share/rhasspy/profiles/en/kaldi/model/data/lang/words.txt',
'--keep_isymbols=false', '--keep_osymbols=false',
'/share/rhasspy/profiles/en/kaldi/language_model.txt',
'/share/rhasspy/profiles/en/kaldi/model/data/lang/G.fst.unsorted']'
returned non-zero exit status 1.



*Taymour*

**Comment by synesthesiam** [comment]
Mixing English and Arabic sentences won't work unfortunately :slightly_frowning_face:

**Comment by B0ndo2** [comment]
So how do I run Arabic ?



*Taymour*

**Comment by dejan2101** [comment]
Hello, what would be needed to add support for Slovenian language?

**Comment by jose1711** [comment]
I would be interested in Slovak language however Mozilla Voice only has 28 hours of recorded voice. Is that number too low to get a meaningful model?
